### PR TITLE
Support _variant in outer level enum formatting for Display

### LINKED ISSUE #142: `affix` is confusing with outer-level `enum` display formatting
Currently, when one uses a top-level `display` attribute for the `Display` derive on an `enum`, it switches to an affix mode that defines the `enum`'s overall format and expects, at most, a single placeholder where inner variant display logic can print into:

```rust
#[derive(Display)]
#[display(fmt = "Prefix {} Suffix")] // at most 1 placeholder allowed here
enum Enum {
    #[display(fmt ="value")] // prints "Prefix value Suffix"
    Variant1,
    // ...
}
```

However, there's a few problems with this current behavior:

* It's REALLY surprising to discover that `fmt` works rather differently than anywhere else here. It's not obvious why formatting _should_ be different here, and there's no "obvious" signal that behavior is different. I feel qualified to say this since _I implemented this behavior myself a year ago_, and still was confused after reading diagnostics until I remembered what I had implemented.
* Affix formatting defined this way is currently all-or-nothing. Once the design of the `Display` impl requires using a specific format some of the time,

I think that we can do better here. There are enough surprises with the current behavior that I think exploring alternatives is warranted. Here's what we stand to gain:

* We can develop an API that operates as similarly to variant-level formatting as possible, which makes the design more coherent.
* Because `derive_more` inhabits the same ecosystem as other crates that do `Display` deriving with rather different approaches, we could possibly try to find a top-level `enum` formatting option that is also consistent with those approaches (and play a bit better with other crates conceptually). In a second, I'll propose something similar to what `thiserror`, an error crate with facilities for deriving `Display`, currently does.

So, here's my proposal:

* The current affix-oriented behavior should be moved to an `affix` attribute key, and (this) new behavior will be defined for `fmt` instead.
* Re-define the `fmt` spec at the top level of an `enum` to be the _default_ printing logic for an `enum`'s variants that can be overridden by a more specific `fmt` rule. An overridable default has basis in some prior art; `thiserror::Error` treats top-level `enum` format specifiers as the "default" message ([motivating issue](https://github.com/dtolnay/thiserror/issues/18), [current relevant upstream `master` code at time of writing](https://github.com/dtolnay/thiserror/blob/04b91d76462834279a323b7771b48158aba19db4/impl/src/ast.rs#L83)):

    ```rust
    use thiserror::Error;
    #[derive(Debug, Error)]
    #[error("error")]
    enum Enum {
        A, // displays "error"
        #[error("b")]
        B, // displays "b"
    }
    ```
* Disallow `affix` and `fmt` to be used together at the same time for the time being, since that's a lot of open design space that doesn't need to be handled right this second.

To concretely translate my above suggestions into code:

```rust
use derive_more::Display;

// Using affixes at the top level of an `enum` is the same as before, except we
// use an `affix` key in the `display` attribute instead of `fmt`.
#[derive(Display)]
#[display(affix = "<{}>")]
enum UsesAffix {
    #[display(fmt = "a")]
    A, // prints "<a>"
    #[display(fmt = "b")]
    B, // prints "<b>"

	// This makes no sense and is still disallowed.
	// #[display(affix = ...)]
	// C,

    // ...
}

// A top-level `display(fmt = ..., ...)` now is:
//
// * Just like variant-level `display` attributes in that it uses formatting
//     args as usual. Only `self` is bound in arguments' context, of course.
// * Overridable by a variant-level `display` attribute.
//
// One motivating example for this is to just use `Debug` printing
// by default unless we happen to have something better.
#[derive(Display)]
#[format(fmt = "{:?}", self)]
enum DisplayDefaultsToDebug {
	A, // prints "A"
	#[display(fmt = "Display B")]
	B, // prints "Display B"
}
```

---

I'm working on a draft PR for this [here](https://github.com/JelteF/derive_more/pull/143). I'm speculatively doing work without getting validation in this issue first, but I don't mind. :)

### LINKED ISSUE #239: ParseError when implement Display derive for enum
Awesome library, i like it!

I found some small problems during use, in some cases, it will be panic. The following is an example:

```rust
#[derive(Display)]
#[display(fmt = "{_0}")]
enum NodeType {
    Base(&'static str),
    Main(&'static str),
    N1(&'static str),
    N2(&'static str),
    N3(&'static str),
    N4(&'static str),
}
```

Here is the compile error result:
```
error: proc-macro derive panicked
  --> src/main.rs:93:10
   |
93 | #[derive(Display)]
   |          ^^^^^^^
   |
   = help: message: called `Result::unwrap()` on an `Err` value: ParseError { line: 1, column: 2, offset: 1, expected: {"}", "[0-9]", ":"} }
```

I can't find the cause of the problem from the message.

### PR DESCRIPTION
**Pull request by JelteF**
Resolves #142
Resolves #239

## Synopsis

This adds back support for top-level format strings of the Display derive. It
now includes the display of the variant whenever the `_variant` placeholder
apears be found. It also supports using the field

## Solution

This does not include the same support for Debug, since it is considered much less
useful there and the differences between the Debug implementation and Display
implementation make it a non-trivial port.

Only named arguments are supported in the format string, no positional ones.
This made the implementation easier, maybe in a future PR positional support
can be added.

This bumps MSRV to 1.70.0, on earlier versions the derived code throws the following error:
```rust
error: there is no argument named `_0`
    --> tests/display.rs:1373:26
     |
1373 |                 #[derive(Display)]
     |                          ^^^^^^^
     |
     = note: did you intend to capture a variable `_0` from the surrounding scope?
     = note: to avoid ambiguity, `format_args!` cannot capture variables when the format string is expanded from a macro
     = note: this error originates in the derive macro `Display` (in Nightly builds, run with -Z macro-backtrace for more info)
 ```

## Checklist

- [x] Documentation is updated (if required)
- [x] Tests are added/updated (if required)
- [x] [CHANGELOG entry][l:1] is added (if required)

[l:1]: /CHANGELOG.md

### CODE REVIEW

**Comment by JelteF** [comment]
Still needs some cleanup, error-ing on edge cases, and updated docs, but the general design seems to work.

**Comment by JelteF** [comment]
> Still needs some cleanup, error-ing on edge cases, and updated docs, but the general design seems to work.

Okay all done now, this is ready for review.

File: clippy.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # See full lints list at:
 # https://rust-lang.github.io/rust-clippy/master/index.html
 
-msrv = "1.65.0"
+msrv = "1.70.0"
```
Comment (tyranron) [inline clippy.toml:5]: I think we can remove it completely, because it defaults to `Cargo.toml`'s `rust-version` at this point.

File: tests/display.rs [hunk 2]
```diff
@@ -1283,6 +1283,131 @@ mod enums {
...
             }
         }
+
+        mod shared_format {
+            use super::*;
+            mod single {
+                use super::*;
+
+                #[derive(Display)]
+                #[display("Variant: {_variant}")]
```
Comment (tyranron) [inline tests/display.rs:1293]: I think there should be also tests like `#[display("Variant: {}", _variant)]`

File: tests/compile_fail/display/shared_format_positional_placeholders.stderr [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+2 | #[display("Stuff({})")]
+  |           ^^^^^^^^^^^
+
+error: shared format string cannot contain positional placeholders, use named placeholders instead
+ --> tests/compile_fail/display/shared_format_positional_placeholders.rs:8:11
+  |
+8 | #[display("Stuff({0})")]
+  |           ^^^^^^^^^^^^
+
+error: shared format string does not support positional placeholders, use named placeholders instead
```
Comment (tyranron) [inline tests/compile_fail/display/shared_format_positional_placeholders.stderr:13]: This seems to me a quite misleading error, because in tests we clearly have the valid case `#[display("Variant {_variant} {_0}")]` where the positional placeholder is used.

**Review by tyranron — CHANGES_REQUESTED** [review]

File: tests/compile_fail/display/shared_format_positional_placeholders.stderr (hunk 3 above)
Comment (JelteF) [reply tests/compile_fail/display/shared_format_positional_placeholders.stderr:13]: That's not a positional placeholder. That's a named placeholder for a tuple field. But I do understand the confusion. If you have a suggestion for better wording of the error, I'm happy to change it.

File: tests/display.rs (hunk 2 above)
Comment (JelteF) [reply tests/display.rs:1289]: the `_0` also doesn't work as a positional placeholder. Adding support for positional placeholders is non-trivial, because the current implementation splits the format string and calls `write!` multiple times. So if you have something like:
```rust
#[display("{}, {_variant}, {}", 123, 456)]
```
Then you need to move the expressions to the correct place.
Because this will result in the wrong output:
```rust
write!("{}, ", 123, 456)
// Write variant
write!(", {}", 123, 456)
```

I agree that positional parameters are useful, but I think we only need named parameters for 1.0. In 0.99 we didn't even support parameters at all, only the variant itself. So even with the current implementation we have a superset of the features of 0.99. And that is really all that I think we should be aiming for for 1.0 at this point. Otherwise the release might take another year.

Comment (tyranron) [reply tests/display.rs:1289]: @JelteF so, as I see, the problem lays in the way the code desugars.

Why not try to leverage `format_args!()` and do as we usually do?
```rust
let _variant = format_args!("whatever variant formatting is", 1, 2);
write!(f, "{}, {_variant}, {}, {}", 123, 456, _variant)
```

Were there any problems with such design?

Comment (tyranron) [reply tests/display.rs:1289]: Ahh... yes... https://github.com/rust-lang/rust/issues/92698  
Let me think...

Comment (tyranron) [reply tests/display.rs:1289]: Okay, this desugaring just works:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5868f867756cdddd4585eac6dff1856b

It should also simplify the implementation, as we don't need all the machinery to cut the expression to separate `write!`s.

Comment (tyranron) [reply tests/display.rs:1289]: If you're short on time with this now, I may pick up and re-implement, if you don't mind.

Comment (JelteF) [reply tests/display.rs:1289]: That desugaring is a lot better for sure. So yes, if you have time now, please re implement this.

**Review by tyranron — CHANGES_REQUESTED** [review]
Needs [desugaring re-implementation](https://github.com/JelteF/derive_more/pull/377#discussion_r1688168385) as was [discussed](https://github.com/JelteF/derive_more/pull/377#discussion_r1688397549).

**Comment by tyranron** [comment]
@JelteF I did the re-implementation, now it seems fine. [This restriction](https://github.com/JelteF/derive_more/pull/377/files#diff-12c6dc13d5c48b38966179cfd43412034bbe33bbbd2f00880484e4b2344f3930R117-R128) is not necessary, actually, and by doing some deeper inspection of the shared formatting attribute we may lift it in the future. This especially may be useful for the use-cases where we want to add some special modifiers (for octals, etc) for all the variants.  I haven't done it to omit complicating this PR and to not slowing the 1.0 release, as it requires more time and rich test suite.

The only thing I want to add before merging this PR is to extend tests suite a little bit more. Will do it shortly.

File: impl/doc/display.md [hunk 4]
```diff
@@ -185,6 +192,13 @@ enum E {
     Path(PathBuf),
 }
 
+#[derive(Display)]
+#[display("Enum E2: {:?}", _0)]
```
Comment (JelteF) [inline impl/doc/display.md:196]: Let's use the modern syntax in the docs

```suggestion
#[display("Enum E2: {_0:?}")]
```

**Review by JelteF — COMMENTED** [review]
Overall soooo much better than my original implementation. Left a few small notes.

File: impl/src/fmt/display.rs [hunk 5]
```diff
@@ -224,70 +242,134 @@ impl<'a> Expansion<'a> {
...
+                            }
+                        }
+                    } else if self.shared_attr.is_some() {
+                        quote! { derive_more::core::format_args!(#fmt) }
+                    } else {
+                        quote! { derive_more::core::write!(__derive_more_f, #fmt) }
+                    }
+                }
+                None if self.fields.is_empty() => {
+                    let ident_str = self.ident.to_string();
```
[the remark this answers is not in the document]
Comment (tyranron) [reply impl/src/fmt/display.rs:272]: Good catch! We have [`syn::ext::IdentExt::unraw()`](https://docs.rs/syn/2.0.72/syn/ext/trait.IdentExt.html#tymethod.unraw) for this!.

File: impl/src/fmt/display.rs [hunk 6]
```diff
@@ -224,70 +242,134 @@ impl<'a> Expansion<'a> {
...
+                            trait_name_to_attribute_name(self.trait_ident),
+                        ),
+                    ))
+                }
+            };
         }
+
+        if let Some(shared_fmt) = &self.shared_attr {
+            let shared_body = if let Some((shared_expr, shared_trait_ident)) =
+                shared_fmt.transparent_call()
```
[the remark this answers is not in the document]
Comment (tyranron) [reply impl/src/fmt/display.rs:316]: @JelteF ahh, yes... that's why I wanted to extend tests suite a little bit more. Good catch, again.

This issue is fixable by additionally referencing the `format_args!()` expression. Cove

> I think it would be best if we just ignored the whole shared_attr logic if the shared attr (both here in and in the big code block above). i.e. if the shared attr is `"{_variant}"` then we consider shared_attr to be `None`.

Yes, I was thinking about it too, but left "as is" as a paved way for future supporting of `{_variant}` with modifiers.

File: impl/src/fmt/display.rs [hunk 7]
```diff
@@ -224,70 +242,134 @@ impl<'a> Expansion<'a> {
...
+            .map_or(true, |a| a.contains_arg("_variant"))
+        {
+            body = match &self.attrs.fmt {
+                Some(fmt) => {
+                    if let Some((expr, trait_ident)) = fmt.transparent_call() {
+                        if self.shared_attr.is_some() {
+                            let placeholder =
+                                trait_name_to_default_placeholder_literal(&trait_ident);
+
+                            quote! { derive_more::core::format_args!(#placeholder, #expr) }
```
[the remark this answers is not in the document]
Comment (tyranron) [reply impl/src/fmt/display.rs:259]: @JelteF 

> I think in my version I was actually still doing a transparent call, even when there was a shared format. It makes sense that that's not possible with your approach, and I think that is fine.

As I've [mentioned before](https://github.com/JelteF/derive_more/pull/377#issuecomment-2248649603), this is possible (to some reasonable degree, at least), by doing some deeper inspection of the shared formatting attributing and by using `&dyn fmt::Trait` for transparent calls. But this is a theme for the follow-up after-1.0-release PR.

> However, I think this code is a bit needlessly complex and by switching up the comparisons a bit it's much much easier to follow:

Makes sense.

File: impl/src/fmt/display.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tyranron) [reply impl/src/fmt/display.rs:316]: However, thinking about it more... one thing does not exclude another one.

**Review by tyranron — DISMISSED** [review]

**Comment by tyranron** [comment]
@JelteF if you don't have any more suggestions on this, I'm ready to merge this one.

File: tests/display.rs [hunk 8]
```diff
@@ -1278,6 +1284,154 @@ mod enums {
...
+
+                #[derive(Display)]
+                #[display("{_variant}")]
+                enum Enum {
+                    #[display("A {_0}")]
+                    A(i32),
+                    #[display("B {}", field)]
+                    B {
+                        field: i32,
+                    },
```
Comment (JelteF) [inline tests/display.rs:1325]: I would want to see a transparent field as well, that correctly passes on the format options. It should 

```
// adding this variant the list
#[display("{:b}")]
Binary(i32)

// and then doing this test
assert_eq!(format("{:08}", Enum::Binary(4)), "00000100");
```

Comment (tyranron) [reply tests/display.rs:1325]: Done.

**Review by tyranron — APPROVED** [review]
